docs/interfaces/cli.md

```diff
@@ -1,3 +1,292 @@
+
+Here is an example of a command a user might enter to interact with the nameservice CLI `nscli` in order to buy a name:
+
+```bash
+nscli tx nameservice buy-name <name> <amount> --gas auto --gas-prices <gasPrices>
+```
+The first four strings specify the command: the root command for the entire application `nscli`, the subcommand `tx`, the subcommand `nameservice` to indicate which module to route the command to, and the type of transaction `buy-name`. The next two strings are arguments: the `name` the user wishes to buy and the `amount` they want to pay for it. Finally, the last few strings of the command are flags to indicate how much the user is willing to pay in fees (calculated using the amount of gas used to execute the transaction and the gas prices provided by the user).
+
+The CLI interacts with a node (running `nsd`) to handle this command.
+
+### Main Function
+
+The `main.go` file needs to have a `main()` function that does the following to run the command-line interface:
+
+* **Instantiate the `codec`** by calling the application's `MakeCodec()` function. The [`codec`](../core/encoding.md) is used to code and encode data structures for the application - stores can only persist `[]byte`s so the developer must define a serialization format for their data structures or use the default, [Amino](../core/amino.md).
+* **Configurations** are set by reading in configuration files (e.g. the sdk config file).
+* **Create the root command** to which all the application commands will be added as subcommands and add any required flags to it, such as `--chain-id`.
+* **Add subcommands** for all the possible user interactions, including [transaction commands](#transaction-commands) and [query commands](#query-commands).
+* **Create an Executor** and execute the root command.
+
+An example of the `main()` function for the [nameservice tutorial](https://cosmos.network/docs/tutorial) CLI can be found [here](https://github.com/cosmos/sdk-application-tutorial/blob/master/cmd/nscli/main.go#L26-L67). The rest of the document will detail what needs to be implemented for each step and include smaller portions of code from the nameservice CLI `main.go` file.
+
+## Commands
+
+Every application CLI first constructs a root command, then adds functionality by aggregating subcommands (often with further nested subcommands) using `AddCommand()`. The bulk of an application's unique capabilities lies in its transaction and query commands, called `TxCmd` and `QueryCmd` respectively.
+
+### Root Command
+
+The root command (called `rootCmd`) is what the user first types into the command line to indicate which application they wish to interact with. The string used to invoke the command (the "Use" field) is typically the name of the application suffixed with `-cli`, e.g. `appcli`. The root command must include the following commands to support basic functionality in the application.
```

Comment:
The root command must not include anything. Best we can say is: "Here are a few root commands that are typically included in SDK applications:"